src/serac/physics/utilities/functional/integral_utilities.hpp

```diff
@@ -230,6 +232,29 @@ struct qf_result<lambda_type, x_t, u_du_t, qpt_data_type, std::enable_if_t<!std:
                                     std::add_lvalue_reference_t<qpt_data_type>>;
 };
 
+/**
+ * @brief derivatives_ptr access
+ *
+ * Templating this will allow us to change the stride-access patterns more consistently
+ *
+ * @param[in] derivative_ptr pointer to derivatives
+ * @param[in] e element number
+ * @param[in] q qaudrature number
+ * @param[in] rule quadrature rule
+ * @param[in] num_elements number of finite elements
+ */
+template <typename derivatives_type, typename rule_type, bool quadrature_coalescing = true>
+SERAC_HOST_DEVICE constexpr derivatives_type& AccessDerivatives(derivatives_type* derivatives_ptr, int e, int q,
+                                                                [[maybe_unused]] rule_type& rule,
+                                                                [[maybe_unused]] int        num_elements)
+{
+  if constexpr (quadrature_coalescing) {
+    return derivatives_ptr[e * int(rule.size()) + q];
+  } else {
+    return derivatives_ptr[q * num_elements + e];
+  }
+}
```

Comment:
we could just accept a stride as the last argument, and not have to have the `[[maybe_unused]]`, e.g.
```cpp
template <typename derivatives_type, typename rule_type, bool quadrature_coalescing = true>
 SERAC_HOST_DEVICE constexpr derivatives_type& AccessDerivatives(derivatives_type* derivatives_ptr, int e, int q, int stride)
 {
   if constexpr (quadrature_coalescing) {
     return derivatives_ptr[e * stride + q];
   } else {
     return derivatives_ptr[q * stride + e];
   }
 }
```

but this will get cleaner when we have a multidimensional array container

Reply:
Maybe with a multidimensional array that will be better? Ideally we can enforce the strides properly the same way you have everything enforced in the `tensor` classes. I'd like to avoid specifying specifying `stride` as an argument since that can lead to error prone behavior.

Reply:
yep, we have to jump through these hoops to index things because we just don't have a proper container! I think mfem::DeviceTensor is pretty close to what we want, except that it's purely a view.